nexus/blueprint-execution/src/omicron_zones.rs

```diff
@@ -0,0 +1,346 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Manges deployment of Omicron zones to Sled Agents
+
+use anyhow::Context;
+use futures::stream;
+use futures::StreamExt;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::lookup::LookupPath;
+use nexus_db_queries::db::DataStore;
+use nexus_types::deployment::OmicronZonesConfig;
+use sled_agent_client::Client as SledAgentClient;
+use slog::info;
+use slog::warn;
+use slog::Logger;
+use std::collections::BTreeMap;
+use uuid::Uuid;
+
+/// Idempotently ensure that the specified Omicron zones are deployed to the
+/// corresponding sleds
+pub async fn deploy_zones(
+    log: &Logger,
+    opctx: &OpContext,
+    datastore: &DataStore,
+    zones: &BTreeMap<Uuid, OmicronZonesConfig>,
+) -> Result<(), Vec<anyhow::Error>> {
+    let errors: Vec<_> = stream::iter(zones)
+        .filter_map(|(sled_id, config)| async move {
+            let client = match sled_client(opctx, datastore, *sled_id).await {
+                Ok(client) => client,
+                Err(err) => {
+                    warn!(log, "{err:#}");
+                    return Some(err);
+                }
+            };
+            let result =
+                client.omicron_zones_put(&config).await.with_context(|| {
+                    format!("Failed to put {config:#?} to sled {sled_id}")
+                });
+
+            match result {
+                Err(error) => {
+                    warn!(log, "{error:#}");
+                    Some(error)
+                }
+                Ok(_) => {
+                    info!(
+                        log,
+                        "Successfully deployed zones for sled agent";
+                        "sled_id" => %sled_id,
+                        "generation" => config.generation.to_string()
```

Comment:
Tiny nit, since `generation` impls `Display`:

```suggestion
                        "generation" => %config.generation,
```

(This might really not matter if `slog` just calls `.to_string()` anyway, but it matches `%sled_id` :man_shrugging:)

Reply:
👍 done in cdc88d0b7.